File: Data Structures & Algorithms/binary-tree-diameter/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    //diameter of a binary tree is the number of edges on the longest path 
    //so use DFS 

    static int maxDiameter = 0; 

    private int dfsHelper(TreeNode node){
        if(node == null) return 0;

        int leftH = dfsHelper(node.left); 
        int rightH = dfsHelper(node.right); 

        if(leftH + rightH > maxDiameter){
            maxDiameter = leftH + rightH; 
        }

        return 1 + Math.max(leftH, rightH); 
    }

    public int diameterOfBinaryTree(TreeNode root) {
        maxDiameter = 0; 
        dfsHelper(root); 

        return maxDiameter;
    }
}

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {  

    private bool IsSame(TreeNode tree1, TreeNode tree2){
        if(tree1 == null && tree2 == null) return true;

        if(tree1 == null || tree2 == null) return false;

        return (tree1.val == tree2.val) 
        && IsSame(tree1.left, tree2.left) 
        && IsSame(tree1.right, tree2.right);
    }  
    public bool IsSubtree(TreeNode root, TreeNode subRoot) {
        if(root == null) return false; 

        if(subRoot == null) return true;

        if(IsSame(root, subRoot)){
            return true;
        }

        return IsSubtree(root.left, subRoot) || IsSubtree(root.right, subRoot);
    }
}
